### DISCUSSION TITLE
[API Proposal]: add yield for Stream

### DISCUSSION THREAD

**Discussion by Mr0N** [dotnet/csharplang#6594] [Language Ideas]
### Background and motivation

I suggest adding an analogue of yield only Stream.

### API Proposal

```csharp
 public Stream Get()
        {
            yield return (byte)1;
            yield return new byte[] {92398,982398,872378}
            yield return Enumerable.Range(0, 1000).Select(a => (byte)a);
        }
```

### API Usage

```csharp
 class Example
    {
        public void Write()
        {
            using var stream = Get();
            using var client = File.OpenWrite("testfile.txt");
            Get().CopyTo(client);
        }
        public Stream Get()
        {
            yield return (byte)1;
            yield return new byte[] { 92398, 982398, 872378 }
            yield return Enumerable.Range(0, 1000).Select(a => (byte)a);
        }
    }

```

**Comment by HaloFour** [comment]
What would the use case be for something like this?

It'd be [pretty trivial](https://stackoverflow.com/questions/22047900/ienumerable-to-stream) to write an implementation of `Stream` that abstracts over an `IEnumerable<byte>` or `IEnumerable<byte[]>` to allow the use of an iterator as the source.

**Reply by Mr0N** [reply]
> What would the use case be for something like this?
> 
> It'd be [pretty trivial](https://stackoverflow.com/questions/22047900/ienumerable-to-stream) to write an implementation of `Stream` that abstracts over an `IEnumerable<byte>` or `IEnumerable<byte[]>` to allow the use of an iterator as the source.

Well, it would be IEnumerable with IEnumerable logic

**Reply by Mr0N** [reply]
listing everything one byte at a time is not a very good idea.

**Reply by HaloFour** [reply]
> Well, it would be IEnumerable with IEnumerable logic

That's the solution, what's the problem that the solution actually intends to solve?  We've had iterators for 17 years and this hasn't come up.

> listing everything one byte at a time is not a very good idea.

An implementation could work with chunks instead of single bytes.

**Reply by Mr0N** [reply]
> An implementation could work with chunks instead of single bytes.

Well, it is already possible in fragments), only it is necessary to customize everything to yield so that it returns yield (byte)1; not immediately but in a fragment).

> That's the solution, what's the problem that the solution actually intends to solve? We've had iterators for 17 years and this hasn't come up.
> 
> >

Well, I had to generate a hash from `IEnumerable<string>`, such a thing would be useful there.

**Reply by HaloFour** [reply]
Those are things an implementation can manage and fairly efficiently.  If such a task was common we'd likely already have one in the BCL, but it doesn't sound like a common enough of a use case to warrant that, let alone a language change, to support it.

**Reply by Mr0N** [reply]
Well, I don't know, I think it's a cool idea)

**Reply by HaloFour** [reply]
Sure, it's a cool idea.  But unless it provides a lot of value it's unlikely to be considered as a language request, especially since it's relatively easy to implement yourself.

**Reply by Mr0N** [reply]
> Sure, it's a cool idea. But unless it provides a lot of value it's unlikely to be considered as a language request, especially since it's relatively easy to implement yourself.

yield can also be implemented yourself.

**Reply by HaloFour** [reply]
> yield can also be implemented yourself.

Yes it can, but manually implementing an iterator is very cumbersome.  Nobody is arguing over the general value of generator coroutines in the language.

The question is whether there is value in having a special flavor of iterators specifically to serve as the source of a `Stream` implementation, especially given that writing a bridge between `IEnumerable<T>` and `Stream` would be relatively easy and only need to be done once.

**Reply by Mr0N** [reply]
> > yield can also be implemented yourself.
> 
> Yes it can, but manually implementing an iterator is very cumbersome. Nobody is arguing over the general value of generator coroutines in the language.
> 
> The question is whether there is value in having a special flavor of iterators specifically to serve as the source of a `Stream` implementation, especially given that writing a bridge between `IEnumerable<T>` and `Stream` would be relatively easy and only need to be done once.

Well, maybe so, but not here, for example, Current should return an array, and it is not so easy to do it, if you make it return one byte, then it is not a problem to implement it.
Second, if Current should return an array, then `return yield new byte[]{} `should return an array, this is not very good. It would be good if yield returned one byte, and Current returned an array.

**Reply by Mr0N** [reply]
or, for example, yield returned the array, and in Current the array somehow changed relative to what Strem needed

**Reply by Mr0N** [reply]
```
    class Test
    {
        public Stream Get(byte[] array)
        {
            yield return (byte)32;
            yield return (byte)32;
            yield return (byte)32;
            yield return (byte)32;
            foreach (var item in array)
            {
                if(item != 32)
                yield return item;
            }
            yield return array;
            yield return  Encoding.UTF8.GetBytes("HI)");
        }
    }
```

**Reply by HaloFour** [reply]
I don't think any of that matters.  Yes, as a language feature it could be more slick to use, but it has to be a _lot_ more than "cool" or "slick" to justify the _enormous_ cost of being a language feature.  A manual implementation of said `Stream` could cover most of those cases based on what kind of iterator it accepts.  There's little point in worrying about trying to optimize a solution that doesn't solve any problems.

**Reply by Mr0N** [reply]
I like this type of code, everything seems simple

**Reply by Mr0N** [reply]
> I don't think any of that matters. Yes, as a language feature it could be more slick to use, but it has to be a _lot_ more than "cool" or "slick" to justify the _enormous_ cost of being a language feature. A manual implementation of said `Stream` could cover most of those cases based on what kind of iterator it accepts. There's little point in worrying about trying to optimize a solution that doesn't solve any problems.

You can manually implement only the case when yield return an array, because it will work to return a byte, but it will be very slow, probably

**Reply by HaloFour** [reply]
> I like this type of code, everything seems simple

Yes, it is simple.  But is it worth the effort to be a language feature?  Almost definitely not.

> You can manually implement only the case when yield return an array, because it will work to return a byte, but it will be very slow, probably

Sure, a naive implementation of such a `Stream` might be slow, and that probably doesn't matter given how often it's likely to be used and in what circumstances.  Other implementations could be smarter and rely on shared buffers or other approaches to be a lot faster.  They might be slightly more difficult to use than a straight iterator, but it's something that you could accomplish today with relatively little effort.

**Comment by bernd5** [comment]
You could write something like this:
```csharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Linq;

var stream = GetGenStream();
while (stream.ReadByte() is var byteVal && byteVal >= 0){
    Console.WriteLine(byteVal);
}

Stream GetGenStream(){
    var enc = System.Text.UTF8Encoding.UTF8;
    var enumerable = Do().SelectMany(s => enc.GetBytes(s));
    
    return new StreamOverEnumerable(enumerable);
    
    static IEnumerable<string> Do(){
        yield return "Hallo";
        yield return "Welt";
    }
}

class StreamOverEnumerable : Stream
{
    private IEnumerator<byte> _enumerator;

    public StreamOverEnumerable(IEnumerable<byte> input) : this(input.GetEnumerator())
    {
    }

    public StreamOverEnumerable(IEnumerator<byte> enumerator)
    {
        _enumerator = enumerator;
    }   

    public override int Read(byte[] buffer, int offset, int count)
    {
        if (_enumerator == null)
        {
            throw new ObjectDisposedException($"The {nameof(StreamOverEnumerable)} has been disposed.");
        }
        
        var totalBytesRead = 0;
        while (totalBytesRead < count && _enumerator.MoveNext())
        {
            buffer[offset+totalBytesRead] = _enumerator.Current;
            totalBytesRead++;
        }

        return totalBytesRead;
    }

    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            Interlocked.Exchange(ref _enumerator, null)?.Dispose();
        }
        base.Dispose(disposing);
    }
        
    public override void Flush()
        => throw WritingNotSupported();

    public override long Seek(long offset, SeekOrigin origin)
        => throw SeekingNotSupported();

    public override void SetLength(long value)
        => throw SeekingNotSupported();

    public override void Write(byte[] buffer, int offset, int count) 
        => throw WritingNotSupported();

    public override bool CanRead => true;
    public override bool CanSeek => false;
    public override bool CanWrite => false;
    public override long Length => throw SeekingNotSupported();
    public override long Position
    {
        get => throw SeekingNotSupported();
        set => throw SeekingNotSupported();
    }
    
    private static Exception WritingNotSupported() =>
        new NotSupportedException("Writing is not supported.");

    private static Exception SeekingNotSupported() =>
        new NotSupportedException("Seeking is not supported.");
}
```

**Reply by Mr0N** [reply]
> You could write something like this:
> 
> ```cs
> using System;
> using System.Collections.Generic;
> using System.IO;
> using System.Threading;
> using System.Linq;
> 
> var stream = GetGenStream();
> while (stream.ReadByte() is var byteVal && byteVal >= 0){
>     Console.WriteLine(byteVal);
> }
> 
> Stream GetGenStream(){
>     var enc = System.Text.UTF8Encoding.UTF8;
>     var enumerable = Do().SelectMany(s => enc.GetBytes(s));
>     
>     return new StreamOverEnumerable(enumerable);
>     
>     static IEnumerable<string> Do(){
>         yield return "Hello";
>         yield return "World";
>     }
> }
> 
> class StreamOverEnumerable : Stream
> {
>     private int _currentIndex;
>     private IEnumerator<byte> _enumerator;
> 
>     public StreamOverEnumerable(IEnumerable<byte> input) : this(input.GetEnumerator())
>     {
>     }
> 
>     public StreamOverEnumerable(IEnumerator<byte> enumerator)
>     {
>         _enumerator = enumerator;
>     }   
> 
>     public override int Read(byte[] buffer, int offset, int count)
>     {
>         if (_enumerator == null)
>         {
>             throw new ObjectDisposedException($"The {nameof(StreamOverEnumerable)} has been disposed.");
>         }
>         
>         var totalBytesRead = 0;
>         while (totalBytesRead < count && _enumerator.MoveNext())
>         {
>             buffer[offset] = _enumerator.Current;
>             totalBytesRead++;
>             _currentIndex++;
>         }
> 
>         return totalBytesRead;
>     }
> 
>     protected override void Dispose(bool disposing)
>     {
>         if (disposing)
>         {
>             Interlocked.Exchange(ref _enumerator, null)?.Dispose();
>         }
>         base.Dispose(disposing);
>     }
>         
>     public override void Flush()
>         => throw WritingNotSupported();
> 
>     public override long Seek(long offset, SeekOrigin origin)
>         => throw SeekingNotSupported();
> 
>     public override void SetLength(long value)
>         => throw SeekingNotSupported();
> 
>     public override void Write(byte[] buffer, int offset, int count) 
>         => throw WritingNotSupported();
> 
>     public override bool CanRead => true;
>     public override bool CanSeek => false;
>     public override bool CanWrite => false;
>     public override long Length => throw SeekingNotSupported();
>     public override long Position
>     {
>         get => throw SeekingNotSupported();
>         set => throw SeekingNotSupported();
>     }
>     
>     private static Exception WritingNotSupported() =>
>         new NotSupportedException("Writing is not supported.");
> 
>     private static Exception SeekingNotSupported() =>
>         new NotSupportedException("Seeking is not supported.");
> }
> ```

and create IEnumerable<byte> lazy, how?

**Reply by Mr0N** [reply]
not byte by byte, but in parts

**Reply by bernd5** [reply]
The stream here is just a thin wrapper around an IEnumerator<byte> -
here the "chunk" handling is done in SelectMany.
Perhaps you can optimize it - but it should not be bad...

**Reply by Mr0N** [reply]
This is a byte-by-byte iteration.

**Reply by Mr0N** [reply]
it is necessary that Current returns the array so that it does not happen like this

**Reply by bernd5** [reply]
in the adapter - that is true. But not in the generator. It produces chunks like "Hello" and "World"....

**Comment by CyrusNajmabadi** [comment]
The language already supports creating both IEnumerables and IAsyncEnumerables simply using `yield` and `await`.  This is the low level language building block that you can use to create higher level abstractions out of.  I don't see anything for the language to do here.

**Reply by Mr0N** [reply]
![image](https://user-images.githubusercontent.com/61237140/197998106-fc83ffbc-54d7-4867-ae87-cbc08e955817.png)

**Reply by jnm2** [reply]
@Mr0N It is a language request, so it does belong here as a discussion. Cyrus is saying that he thinks this language feature shouldn't be added, because given the existing language features which already cater to this scenario well enough in his opinion, there shouldn't be anything for the language to do.

**Reply by Mr0N** [reply]
What are these?

**Reply by jnm2** [reply]
These show the process of 
.NET team members in moving this thread to the appropriate place.

**Reply by HaloFour** [reply]
A suggestion to the runtime team might be to include an implementation of `Stream` in the BCL that can read from an `IEnumerable<T>` as a source.

**Reply by Mr0N** [reply]
> A suggestion to the runtime team might be to include an implementation of `Stream` in the BCL that can read from an `IEnumerable<T>` as a source.

Well, you can, for example, with IEnumerable<byte[]> or `IEnumerable<IEnumerable<byte>>`
